ENH: update pivot op

Hank0626 opened the pull request:
## What do these changes do?
- support one chunk simplification
- add shape infer and combine op

<!-- Please give a short brief about these changes. -->

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

UranusSeven commented:
Hi, could you provide the reasons behind these changes? Like what is the problem you are trying to solve?

Hank0626 commented:
> Hi, could you provide the reasons behind these changes? Like what is the problem you are trying to solve?

1. Like pivot_table op, pivot will have map, reduce and shuffle operations during computation. After the shuffle operation, it is possible that the first chunk does not contain any data (thus has the wrong shape), but during the test, when judging the shape, we use the shape of the first chunk to infer the shape of the result. So I added yield for inferring the shape and combine to fuse the final result with the ending.
2. I added the function `tile_one_chunk` to deal with the case of only one chunk, so that it can be more efficient to deal with the case of no need to split the chunk

aresnow1 commented:
Optimization is possible here. After yield, you can check if the chunk's columns are the same as the `output_columns`. If so, you can skip the combine phase, which can reduce some calculations.

Hank0626 commented:
> Optimization is possible here. After yield, you can check if the chunk's columns are the same as the `output_columns`. If so, you can skip the combine phase, which can reduce some calculations.

done, also done for pivot_table https://github.com/xprobe-inc/xorbits/pull/485

aresnow1 commented on python/xorbits/_mars/dataframe/base/tests/test_base_execution.py:
```diff
@@ -698,40 +698,48 @@ def sorted_df(df):
     # test basic pivot
     r = df.pivot(index="foo", columns="bar", values="baz")
-    result = r.execute().fetch()
+    result = r.execute(extra_config={"check_dtypes": False}).fetch()
```
Why skip checking dtypes?

Hank0626 replied:
deal with the order of dtypes

aresnow1 reviewed (approved):
LGTM

UranusSeven reviewed (approved):
LGTM. Good job :)
